src/compiler/cse.rs

```diff
@@ -0,0 +1,452 @@
+                    step_type_with_hash.add_internal(*signal);
+                }
+                step_type_with_hash.auto_signals.insert(*q, expr.clone());
+                step_type_with_hash.add_constr(format!("{:?}", q), expr.clone());
+            });
+            decomp.constrs.iter().for_each(|expr| {
+                step_type_with_hash.add_constr(format!("{:?}", expr), expr.clone());
+            });
+
+            // Replace the common subexpression in all constraints
+            step_type_with_hash
+                .constraints
+                .iter_mut()
+                .for_each(|constraint| {
+                    constraint.expr = replace_expr(&constraint.expr, &common_se);
+                });
+        } else {
+            // No more common subexpressions found, exit the loop
+            break;
+        }
+        *step_type = step_type_with_hash.transform_meta(|_| ());
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(super) struct SubexprInfo {
+    count: usize,
+    degree: usize,
+}
+
+/// Information about a subexpression to help find the optimal subexpression to replace.
+impl SubexprInfo {
+    fn new(count: usize, degree: usize) -> Self {
+        Self { count, degree }
+    }
+
+    fn update(&mut self, degree: usize) {
+        self.count += 1;
+        self.degree = self.degree.max(degree);
+    }
+}
+
+/// Find the optimal subexpression to replace in a list of expressions.
+fn find_optimal_subexpression<F: Field + Hash, S: Scoring<F>>(
+    exprs: &[Expr<F, Queriable<F>, HashResult>],
+    replaced_hashes: &HashSet<u64>,
+    scorer: &S,
+) -> Option<Expr<F, Queriable<F>, HashResult>> {
+    let mut count_map = HashMap::<u64, SubexprInfo>::new();
+    let mut hash_to_expr = HashMap::<u64, Expr<F, Queriable<F>, HashResult>>::new();
+
+    // Extract all subexpressions and count them
+    for expr in exprs.iter() {
+        count_subexpressions(expr, &mut count_map, &mut hash_to_expr, replaced_hashes);
+    }
+
+    // Find the best common subexpression to replace based on the score
+    let best_subexpr = count_map
+        .iter()
+        .map(|(&hash, info)| {
+            let expr = hash_to_expr.get(&hash).unwrap();
+            let score = scorer.score(expr, info);
+            (hash, score)
+        })
+        .filter(|&(_, score)| score > 0)
+        .max_by_key(|&(_, score)| score)
+        .map(|(hash, _)| hash);
+
+    best_subexpr.and_then(|hash| hash_to_expr.get(&hash).cloned())
+}
+
+/// Count the subexpressions in an expression and store them in a map.
+fn count_subexpressions<F: Field + Hash>(
+    expr: &Expr<F, Queriable<F>, HashResult>,
+    count_map: &mut HashMap<u64, SubexprInfo>,
+    hash_to_expr: &mut HashMap<u64, Expr<F, Queriable<F>, HashResult>>,
+    replaced_hashes: &HashSet<u64>,
+) {
+    let degree = expr.degree();
+    let hash_result = expr.meta().hash;
+
+    // Only count and store if not already replaced
+    if !replaced_hashes.contains(&hash_result) {
```

Comment:
@rutefig if it is already replaced, it should always count to one, so there is no need to special case